common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java

```diff
@@ -835,6 +835,28 @@ public int indexOf(UTF8String v, int start) {
     return -1;
   }
 
+  public int indexOf(UTF8String substring, int start, int collationId) {
+    if (CollationFactory.fetchCollation(collationId).supportsBinaryEquality) {
+      return this.indexOf(substring, start);
+    }
+    if (collationId == CollationFactory.UTF8_BINARY_LCASE_COLLATION_ID) {
+      return this.toLowerCase().indexOf(substring.toLowerCase(), start);
+    }
+    return collatedIndexOf(substring, start, collationId);
+  }
+
+  private int collatedIndexOf(UTF8String substring, int start, int collationId) {
```

Comment:
Let's go with this naming: `collationAwareIndexOf`

ref: https://github.com/apache/spark/pull/45749#discussion_r1546421174